sdk/clusterplugin/plugin.go

```diff
@@ -0,0 +1,83 @@
+package clusterplugin
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/c3os-io/c3os/sdk/bus"
+	"github.com/mudler/go-pluggable"
+	yip "github.com/mudler/yip/pkg/schema"
+	"gopkg.in/yaml.v2"
+)
+
+const clusterProviderCloudConfigFile = "/usr/local/cloud-config/cluster.c3os.yaml"
+
+// ClusterProvider returns a yip configuration that configures a Kubernetes engine.  The yip config may use any elemental
+// stages after initramfs.
+type ClusterProvider func(cluster Cluster) yip.YipConfig
+
+// ClusterPlugin creates a cluster plugin from a `ClusterProvider`.  It calls the cluster provider at the appropriate events
+// and ensures it configuration is written where it will be executed.
+type ClusterPlugin struct {
+	Provider ClusterProvider
+}
+
+func (p ClusterPlugin) onBoot(event *pluggable.Event) pluggable.EventResponse {
```

Comment:
a question here, it looks this code is not really used - i'm not sure if it's because there are plans for follow-ups. or is that a leftover? maybe this piece doesn't need to sit on c3os

Reply:
on a side note: feel to add unit tests?

Reply:
I was thinking we would build out a few plugin bases like this to encourage best practices with plugins, e.g. this is the preferred way to write cluster plugins but the framework is still flexible enough that users can write their own logic on the events if they like.  This sdk would be to c3os what kube builder is to kubernetes.

Reply:
Sounds good to me :+1:

Reply:
see https://github.com/c3os-io/provider-k3s for an example